Spine-project/Spine-Toolbox
# Issue 1047 conda env activation

Support for Conda environments feature is now ready for testing. This branch should be merged to master alongside a PR in Spine Items and another one in Spine Engine.

To test, checkout branches:
SpineToolbox / issue_1047_conda_env_activation
SpineItems / issue_1047_support_conda_environments
SpineEngine / issue_1047_activate_conda_envs
SpineDbAPI / master (no changes here)

1. Open Spine Toolbox and a project with some Tool specs (Python and others). See that they run like previously.
2. Open a Tool spec editor for some pre-made Python Tool spec. Change something in the new execution settings box and save
3. Run the project again and see that the execution now happens according to your changes
4. Open the Settings widget and go to page Tools
5. Set a Conda executable into the new line edit in the bottom. (i.e *your conda install path*/Scripts/conda.exe). Click Ok to save new Settings.
6. Outside Spine Toolbox, open a conda shell and create a new environment called `abc` for example
7. install ipykernel to this new conda environment (pip install ipykernel)
8. Back in Spine Toolbox, Create a new Tool Spec. Select Python as the Tooltype, give it some name and a hello world script. Select `Jupyter Console` radiobutton. Click the kernel spec combobox, There should be a list item `conda-env-????-abc-py [Conda]` in the drop-down list. The name depends on where your conda envs are located. Select it, save and close.
9. Drag the new Tool spec to Design View and execute. Type in `!conda info` to the Jupyter Console after execution to confirm the active environment.
 
Known issues:
- Loading the Conda kernel specs takes about a 1-2 seconds on my system, so there's a noticeable lag when opening a Python Tool spec editor or when changing the Tool type in Tool spec editor e.g. from Julia to Python. This can be optimized later.
- Tests do not pass on GitHub, because the GitHub action does not install the correct branches from SpineItems and SpineEngine. I'll check that they pass locally before merging.

TODO:
* [x] Revert requirements.txt before merging
